internal/impl/postgresql/input_postgrecdc.go

```diff
@@ -0,0 +1,566 @@
+		var latestOffset *int64
+
+		for {
+			select {
+			case <-nextTimedBatchChan:
+				nextTimedBatchChan = nil
+				flushedBatch, err := batchPolicy.Flush(ctx)
+				if err != nil {
+					p.mgr.Logger().Debugf("Timed flush batch error: %w", err)
+					break
+				}
+
+				if !flushBatch(ctx, p.msgChan, flushedBatch, latestOffset, nil) {
+					break
+				}
+
+				// TrxCommit LSN must be acked when all the messages in the batch are processed
+			case trxCommitLsn, open := <-p.pglogicalStream.AckTxChan():
+				if !open {
+					break
+				}
+
+				p.cMut.Lock()
+				p.cMut.Unlock()
+
+				flushedBatch, err := batchPolicy.Flush(ctx)
+				if err != nil {
+					p.mgr.Logger().Debugf("Flush batch error: %w", err)
+					break
+				}
+
+				callbackChan := make(chan bool)
+				if !flushBatch(ctx, p.msgChan, flushedBatch, latestOffset, &callbackChan) {
+					break
+				}
+
+				<-callbackChan
+				if err = p.pglogicalStream.AckLSN(trxCommitLsn); err != nil {
+					p.mgr.Logger().Errorf("Failed to ack LSN: %v", err)
+					break
+				}
+
+				p.pglogicalStream.ConsumedCallback() <- true
+
+			case message, open := <-p.pglogicalStream.Messages():
+				if !open {
+					break
+				}
+				var (
+					mb  []byte
+					err error
+				)
+				if message.Lsn != nil {
+					parsedLSN, err := LSNToInt64(*message.Lsn)
+					if err != nil {
+						p.logger.Errorf("Failed to parse LSN: %v", err)
+						break
+					}
+					latestOffset = &parsedLSN
+				}
+				if mb, err = json.Marshal(message); err != nil {
+					break
+				}
+
+				batchMsg := service.NewMessage(mb)
+
+				streaming := strconv.FormatBool(message.IsStreaming)
+				batchMsg.MetaSet("streaming", streaming)
+				batchMsg.MetaSet("table", message.Changes[0].Table)
+				batchMsg.MetaSet("operation", message.Changes[0].Operation)
+				if message.Changes[0].TableSnapshotProgress != nil {
+					p.snapshotMetrics.SetFloat64(*message.Changes[0].TableSnapshotProgress, message.Changes[0].Table)
+				}
+				if message.WALLagBytes != nil {
+					p.replicationLag.Set(*message.WALLagBytes)
+				}
+
+				if batchPolicy.Add(batchMsg) {
+					nextTimedBatchChan = nil
+					flushedBatch, err := batchPolicy.Flush(ctx)
+					if err != nil {
+						p.mgr.Logger().Debugf("Flush batch error: %w", err)
+						break
+					}
+					if message.IsStreaming {
+						callbackChan := make(chan bool)
+						if !flushBatch(ctx, p.msgChan, flushedBatch, latestOffset, &callbackChan) {
+							break
+						}
+						<-callbackChan
+					} else {
+						if !flushBatch(ctx, p.msgChan, flushedBatch, latestOffset, nil) {
+							break
+						}
+					}
+
+				}
+			case <-ctx.Done():
+				if err = p.pglogicalStream.Stop(); err != nil {
+					p.logger.Errorf("Failed to stop pglogical stream: %v", err)
+				}
+			}
+		}
+	}()
+
+	return err
+}
+
+func (p *pgStreamInput) asyncCheckpointer() func(context.Context, chan<- asyncMessage, service.MessageBatch, *int64, *chan bool) bool {
```

Comment:
Since this `asyncCheckpointer` is only used as the `flushBatch` func in `Connect()` which is always preceded by `callbackChan := make(chan bool)` and then followed by `<-callbackChan`, I wonder if we can get rid of this `callbackChan`.

Reply:
Sure, but it can be cleaned up a bit. The caller shouldn't have to manage that channel. Here's what I had in mind: https://github.com/redpanda-data/connect/commit/4e05248557f30c5df6b2b0ffd7cfc37a91612d41. I'd have to spend more time on it to see if it can be cleaned up more, but this should be a bit more readable I think.